src/imputors/knn.jl

```diff
@@ -0,0 +1,79 @@
+"""
+    KNN <: Imputor
+
+Imputation using k-Nearest Neighbor algorithm.
+
+# Keyword Arguments
+* `num_nn::Int`: number of nearest neighbors
+* `dist::MinkowskiMetric`: distance metric suppports by `NearestNeighbors.jl` (Euclidean, Chebyshev, Minkowski and Cityblock)
+* `on_complete::Function`: a function to run when imputation is complete
+
+# Reference
+* Troyanskaya, Olga, et al. "Missing value estimation methods for DNA microarrays." Bioinformatics 17.6 (2001): 520-525.
+"""
+# TODO : Support Categorical Distance (NearestNeighbors.jl support needed)
+struct KNN{M} <: Imputor where M <: NearestNeighbors.MinkowskiMetric
+    num_nn::Int
+    dist::M
+    context::AbstractContext
+end
+
+function KNN(; num_nn=1,
+    dist=Euclidean(), context=Context())
+
+    KNN(num_nn, dist, context)
+end
+
+"""
+    impute!(imp::KNN, data::AbstractMatrix)
+
+data : N x D matrix
+"""
+function impute!(data::AbstractMatrix{<:Union{T, Missing}},
+    imp::KNN) where T<:Real
+
+    imp.context() do ctx
+        # Get our before and after views of our missing and non-missing data
+        mmask = ismissing.(data)
+        omask = .!mmask
+
+        mdata = data[mmask]
+        odata = data[omask]
+
+        # Fill in the original data as mean value
+        # TODO : pass Fill in constructor
+        impute!(data, Fill(; value=mean, context=ctx))
+
+        # transpose to D x N for KDTree
+        dataT = float.(collect(transpose(data)))
```

Comment:
I'd probably rename `dataT` to just `transposed`. We also try to avoid camel case to follow our style guide https://github.com/invenia/BlueStyle